plugins/GCodeReader/GCodeReader.py

```diff
@@ -0,0 +1,303 @@
+# Copyright (c) 2016 Aleph Objects, Inc.
+# Cura is released under the terms of the AGPLv3 or higher.
+
+from UM.Application import Application
+from UM.Logger import Logger
+from UM.Math.AxisAlignedBox import AxisAlignedBox
+from UM.Math.Vector import Vector
+from UM.Mesh.MeshReader import MeshReader
+from UM.Message import Message
+from UM.Scene.SceneNode import SceneNode
+from UM.i18n import i18nCatalog
+
+catalog = i18nCatalog("cura")
+
+
+from cura import LayerDataBuilder
+from cura import LayerDataDecorator
+from cura.LayerPolygon import LayerPolygon
+from cura.GCodeListDecorator import GCodeListDecorator
+
+import numpy
+import math
+import re
+from collections import namedtuple
+
+
+# Class for loading and parsing G-code files
+class GCodeReader(MeshReader):
+    def __init__(self):
+        super(GCodeReader, self).__init__()
+        self._supported_extensions = [".gcode", ".g"]
+        Application.getInstance().hideMessageSignal.connect(self._onHideMessage)
+        self._cancelled = False
+        self._message = None
+        self._clearValues()
+        self._scene_node = None
+        self._position = namedtuple('Position', ['x', 'y', 'z', 'e'])
+
+    def _clearValues(self):
+        self._extruder = 0
+        self._layer_type = LayerPolygon.Inset0Type
+        self._layer = 0
+        self._previous_z = 0
+        self._layer_data_builder = LayerDataBuilder.LayerDataBuilder()
+        self._center_is_zero = False
+
+    @staticmethod
+    def _getValue(line, code):
+        n = line.find(code)
+        if n < 0:
+            return None
+        n += len(code)
+        pattern = re.compile("[;\s]")
+        match = pattern.search(line, n)
+        m = match.start() if match is not None else -1
+        try:
+            if m < 0:
+                return line[n:]
+            return line[n:m]
+        except:
+            return None
+
+    def _getInt(self, line, code):
+        value = self._getValue(line, code)
+        try:
+            return int(value)
+        except:
+            return None
+
+    def _getFloat(self, line, code):
+        value = self._getValue(line, code)
+        try:
+            return float(value)
+        except:
+            return None
+
+    def _onHideMessage(self, message):
```

Comment:
User-interaction wise, it is not really nice to abort the loading when closing the message, since it is not very explicit. We have support for explicit actions on Messages, so it is better to add an explicit "Cancel" action to the message and not show the close button.